feat/added-restore-command

ryu-sato opened the pull request:
PR & merge タスク
- https://weseek.myjetbrains.com/youtrack/issue/INFRA-169

refs INFRA-151:  S3 からバックアップファイルの DL と MongoDB リストアする機能を実装する

ryu-sato committed:
refs #151:  S3 からバックアップファイルの DL と MongoDB リストアする機能を実装する

skomma commented on README.md:
```diff
@@ -19,6 +19,7 @@ docker run --rm \
   -e AWS_SECRET_ACCESS_KEY=<Your IAM Secret Access Key> \
   -e S3_TARGET_BUCKET_URL=<Target S3 Bucket URL (s3://...)> \
   [ -e BACKUP_FILE_PREFIX=<Prefix of Backup Filename (default: "backup") \ ]
+  [ -e BACKUPEDDAY=<Day which file is backuped specified in format yyyymmdd (default: <TODAY>) \ ]
```
ここはデフォルトオプション(backup, prune, list)するときの説明だから、ここに追記すべきじゃないと思う。
restore 用の手順は別に書くべき。

skomma replied:
あと、変数名がすっごくわかりづらい。

ryu-sato replied:
restore の説明項目は分けます。
元々の説明文章が README.md 内のデフォルトオプションだと言うことって書いてたっけ？
書かなくていいなら構わないけど気になった。

変数名分かりづらいか。restore したいファイルは PREFIX も何もなくてファイル名をそのまま指定させるようにしてみることにします。

skomma replied:
デフォルトオプションは今後書くべきだと思うけど、現時点では最低限の内容しか README.md には書いてない。

skomma commented on bin/functions.sh:
```diff
@@ -3,6 +3,7 @@ AWSCLI="/usr/bin/aws"
 AWSCLI_PUT_OPT="s3 cp"
 AWSCLI_LIST_OPT="s3 ls"
 AWSCLI_DEL_OPT="s3 rm"
+AWSCLI_PULL_OPT="s3 cp"
```
PUT と PULL 一緒なんだったら COPY_OPT とかにしてしまえば？

ryu-sato replied:
backup も restore も依存してしまうのは避けた方がいいと思ったけど強いこだわりでもないし、今回はDRYに従おうかな。

skomma replied:
今後 awscli 以外を使うことはあまり考えられないからいいんじゃないかなぁ。
Amazon 公式だし。

skomma commented on bin/functions.sh:
```diff
@@ -36,6 +37,14 @@ s3_put_file() {
+# Push the specified file.
+# arguments: 1. target s3 url (s3://...)
+#            2. local filename
+s3_pull_file() {
```
こっちも、put/pull がっちゃんこして s3_copy_file でいいと思う。

ryu-sato replied:
修正します。

skomma commented on bin/restore.sh:
```diff
@@ -0,0 +1,64 @@
+  MONGORESTORE_OPTS="${MONGORESTORE_OPTS} --authenticationDatabase ${MONGODB_AUTHDB}"
+fi
+echo "restore MongoDB..."
+mongorestore -h ${MONGODB_HOST} --drop -v ${TARGET} ${MONGORESTORE_OPTS}
```
drop するかどうかもオプション指定できたらいいんじゃないかなぁ？

ryu-sato replied:
なるほど。mongorestore とデフォルトの動作が違うのは確かに違和感あるので、
* デフォルトは --drop なし
* オプションで --drop 指定可能
にします。

ryu-sato committed:
refs INFRA-151:  S3 からバックアップファイルの DL と MongoDB リストアする機能を実装する
- restore コマンドの使い方を default 動作と分けて README へ記載
- bin/function.sh の AWSCLI_PUT_OPT と AWSCLI_PULL_OPT を AWSCLI_COPY_OPT としてマージ
- bin/function.sh の s3_put_file(), s3_pull_file() を s3_copy_file() としてマージ。合わせて bin/backup.sh, bin/restore.sh から該当関数の呼び出し箇所を修正。
- タブ数などの表記ルールを統一するために .editorconfig を追加
- restore コマンドのオプション引数とそれに応じた動作を変更

skomma commented on README.md:
```diff
@@ -28,3 +28,23 @@ docker run --rm \
 
 and after running this, `backup-YYYYMMdd.tar.bz2` will be placed on Target S3 Bucket.
+
+# How to restore
```
他と揃えて、
```
How to restore
--------------
```
がいいかな。細かいけど。

ryu-sato replied:
Usageの1段落下として記載したつもりでしたが、意図した結果と違ったので直します。

skomma commented on README.md:
```diff
@@ -28,3 +28,23 @@ docker run --rm \
+
+# How to restore
+
+You can use restore command to restore database from backup file.
```
You can use **"restore"** command の方がいいかな。

ryu-sato replied:
修正します。

skomma commented on bin/restore.sh:
```diff
@@ -0,0 +1,67 @@
+echo "restore MongoDB..."
+mongorestore -h ${MONGODB_HOST} -v ${TARGET} ${MONGORESTORE_OPTS}
+
+# delete tarball if restore was successfully over
```
これコンテナが消えた時にファイルも消えるからいらんね。

ryu-sato replied:
backup.sh の方では掃除していた（と思っていたが間違い）し、
コンテナを使いまわした時には一応掃除した方が良いかなと思って記載してました。
backup.sh の方では tar を削除するコードはコメントアウトされているようなので、
restore.sh も展開したファイルはコンテナ削除と共に消えるので対応不要としておきます。

ryu-sato committed:
refs INFRA-151: S3 からバックアップファイルの DL と MongoDB リストアする機能を実装する
- restore.sh 終了時に docker container 削除されれば、バックアップ用 tar ball の展開後ファイルは削除不要のため、該当箇所を削除
- backup.sh 終了時に backup した tar ball を削除する箇所は不要であるとコメントアウトされており、function.sh の delete_localfile_if_exists_on_s3() 関数は利用していなかったため双方から削除
- READEME.md の 'How to restore' のインデントレベルを調整（How to use の 1 レベル下）
